From a54739043c1a2bf57d040d4344e18df39e06acf3 Mon Sep 17 00:00:00 2001
From: Dinesh Karthikeyan <quic_dinek@quicinc.com>
Date: Thu, 31 Mar 2022 12:47:22 +0530
Subject: [PATCH] ath12k: Add qdss full coredump collection support

Add qdss dump collection support in coredump when firmware crash
is simulated.

Signed-off-by: Dinesh Karthikeyan <quic_dinek@quicinc.com>
---
 drivers/net/wireless/ath/ath12k/coredump.c | 13 ++++++++++++-
 1 file changed, 12 insertions(+), 1 deletion(-)

--- a/drivers/net/wireless/ath/ath12k/coredump.c
+++ b/drivers/net/wireless/ath/ath12k/coredump.c
@@ -152,7 +152,7 @@ void ath12k_coredump_download_rddm(struc
 	struct mhi_controller *mhi_ctrl = ar_pci->mhi_ctrl;
 	struct image_info *rddm_img, *fw_img;
 	struct ath12k_dump_segment *segment, *seg_info;
-	int i, rem_seg_cnt = 0, len, num_seg, seg_sz;
+	int i, rem_seg_cnt = 0, len, num_seg, seg_sz, qdss_seg_cnt = 1;
 	int skip_count = 0;
 
 	ath12k_mhi_coredump(mhi_ctrl, false);
@@ -168,6 +168,9 @@ void ath12k_coredump_download_rddm(struc
 	}
 
 	num_seg = fw_img->entries + rddm_img->entries + rem_seg_cnt;
+	if (ab->is_qdss_tracing)
+		num_seg += qdss_seg_cnt;
+
 	len = num_seg * sizeof(*segment);
 
 	segment = kzalloc(len, GFP_KERNEL);
@@ -216,6 +219,14 @@ void ath12k_coredump_download_rddm(struc
 		seg_info++;
 	}
 
+	if (ab->is_qdss_tracing) {
+		seg_info->len = ab->qmi.qdss_mem[0].size;
+		seg_info->addr = ab->qmi.qdss_mem[0].paddr;
+		seg_info->vaddr = ab->qmi.qdss_mem[0].v.ioaddr;
+		seg_info->type = ATH12K_FW_QDSS_DATA;
+		seg_info++;
+	}
+
 	num_seg = num_seg - skip_count;
 	if(!ab->fw_recovery_support) {
 		ath12k_coredump_seg_info.chip_id = ar_pci->dev_id;
